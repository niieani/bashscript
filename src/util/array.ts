/**
 * Combines two arrays by alternating their indices
 * combineAlternate([a, b, c], [1, 2, 3]) === [a, 1, b, 2, c, 3]
 */
export const combineAlternate = <T, Y>(
  arr1: Array<T>,
  arr2: Array<Y>,
  noLengthCheck = false,
  flipOrder = false,
): Array<T | Y> =>
  noLengthCheck || arr1.length >= arr2.length
    ? arr1.reduce(
        (combined, arr1value, index) =>
          index < arr2.length
            ? [
                ...combined,
                ...(flipOrder
                  ? [arr2[index], arr1value]
                  : [arr1value, arr2[index]]),
              ]
            : [...combined, arr1value],
        [] as Array<T | Y>,
      )
    : combineAlternate(arr2, arr1, true, true)

export const findLastIndex = <T>(
  predicate: (value: T, index: number, obj: T[]) => boolean,
) => (arr: ReadonlyArray<T>) => {
  const result = arr
    .slice()
    .reverse()
    .findIndex(predicate)
  return result >= 0 ? arr.length - 1 - result : result
}

export const last = <T>(array: Array<T>): T | undefined =>
  array.slice().reverse()[0]

export const ensureArray = <T>(maybeArray: T | Array<T>): Array<T> =>
  Array.isArray(maybeArray) ? maybeArray : [maybeArray]
