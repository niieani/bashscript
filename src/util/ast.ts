import * as AST from 'ts-simple-ast'
import * as acorn from 'acorn'
import {comment} from '../writer/syntax/comment'

export const getCommentObjects = (node: AST.Node) =>
  getComments(node).map((text) => comment(text, false))

export const getComments = ({compilerNode}: AST.Node) => {
  const commentPart = compilerNode
    .getFullText()
    .slice(0, compilerNode.getLeadingTriviaWidth())
  const comments: Array<acorn.Comment> = []
  acorn.parse(commentPart, {
    ranges: false,
    onComment: comments,
    allowHashBang: true,
  })
  return comments.map((comment) => `${comment.value.split('\n').join('\n#')}`)
}
