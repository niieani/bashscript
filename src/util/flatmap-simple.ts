export const flatmapSimple = <T>(nestedArray: Array<Array<T>>): Array<T> =>
  ([] as Array<T>).concat(...nestedArray)
