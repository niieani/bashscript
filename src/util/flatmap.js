// from https://gist.github.com/Kovensky/921c58de0c58be01aebf06d86cf9692f

if (typeof Array.prototype.flatMap !== 'function') {
  // eslint-disable-next-line no-extend-native
  Object.defineProperty(Array.prototype, 'flatMap', {
    enumerable: false,
    // https://bterlson.github.io/proposal-flatMap/#sec-Array.prototype.flatMap
    value: function flatMap (callback, thisArg = undefined) {
      const O = toObject(this)
      const A = arraySpeciesCreate(O, 0)
      // typo in the github pages render, it's missing
      // the ", 1" argument, but it is present in
      // the master source code
      flattenIntoArray(A, O, 0, 1, callback, thisArg)
      return A
    }
  })
}

if (typeof Array.prototype.flatten !== 'function') {
  // eslint-disable-next-line no-extend-native
  Object.defineProperty(Array.prototype, 'flatten', {
    enumerable: false,
    // https://bterlson.github.io/proposal-flatMap/#sec-Array.prototype.flatten
    value: function flatten (depth = Infinity) {
      const O = toObject(this)
      const A = arraySpeciesCreate(O, 0)
      depth = Number(depth)
      flattenIntoArray(A, O, 0, depth)
      return A
    }
  })
}

// https://bterlson.github.io/proposal-flatMap/#sec-FlattenIntoArray
function flattenIntoArray (target, source, start, depth, mapper, thisArg) {
  let targetIndex = start
  let sourceIndex = 0
  const sourceLen = toLength(source.length)

  while (sourceIndex < sourceLen) {
    const P = `${sourceIndex}`
    if (P in source) {
      let element = source[P]
      if (mapper) {
        element = mapper.call(thisArg, element, sourceIndex, target)
      }
      let spreadable
      // https://tc39.github.io/ecma262/#sec-isconcatspreadable
      if (typeof element !== 'object') {
        spreadable = false
      } else {
        spreadable = element[Symbol.isConcatSpreadable]
        if (typeof spreadable !== 'undefined') {
          spreadable = !!spreadable
        } else {
          spreadable = Array.isArray(element)
        }
      }

      // spec deviation: the spec specifies "depth >= 0",
      // but that incorrectly flattens one level too far.
      // https://github.com/bterlson/proposal-flatMap/issues/11
      if (spreadable && depth > 0) {
        const nextIndex = flattenIntoArray(target, element, targetIndex, depth - 1)
        // spec deviation: decrement targetIndex by 1 here
        // it will get incremented back at the tail of the loop
        // https://github.com/bterlson/proposal-flatMap/issues/13
        targetIndex = nextIndex - 1
      } else {
        if (targetIndex !== toLength(targetIndex)) {
          throw TypeError()
        }
        target[targetIndex] = element
      }
    }
    targetIndex += 1
    sourceIndex += 1
  }
  return targetIndex
}

// https://tc39.github.io/ecma262/#sec-arrayspeciescreate
function arraySpeciesCreate (original, length) {
  length = toLength(length)
  if (length < 0) {
    throw new RangeError()
  }

  let C
  if (!Array.isArray(original)) {
    C = original.constructor

    // Trick to implement the Realm check
    // from core-js
    if (typeof C === 'function' &&
      (C === Array || Array.isArray(C.prototype))) {
      C = undefined
    }

    if (typeof C === 'object' ? C !== null : typeof C === 'function') {
      C = C[Symbol.species]
      if (C === null) {
        C = undefined
      }
    }
  }

  if (typeof C === 'undefined') {
    C = Array
  }

  return new C(length)
}

// These are more-or-less like the core-js implementations
// in modules/_to-{length,integer,object}

function toLength (arg) {
  return arg > 0 ? Math.min(toInteger(arg), 0x1fffffffffffff) : 0
}

function toInteger (arg) {
  arg = +arg
  return isNaN(arg) ? 0 : (arg > 0 ? Math.floor : Math.ceil)(arg)
}

function toObject (arg) {
  if (arg === undefined) {
    throw new TypeError(`Can't call method on ${arg}`)
  }
  return Object(arg)
}
