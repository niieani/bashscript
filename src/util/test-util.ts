import {Project} from 'ts-morph'

export const makeTestFile = (fileContents: string) => {
  const project = new Project({
    useInMemoryFileSystem: true,
    addFilesFromTsConfig: false,
    compilerOptions: {
      strict: true,
      noLib: true,
    },
  })
  return project.createSourceFile('main.ts', fileContents)
}
