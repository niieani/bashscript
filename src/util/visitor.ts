import {ScopeContext} from '../writer/types'

export const passthrough = <T>(t: T) => t

export const getNameGenerator = (name: string) => <T>(
  wrapper: (t: string) => T = passthrough as (t: string) => T,
) => ({_context: {scopePath}}: ScopeContext) =>
  wrapper(`${scopePath.join('.')}.${name}`)
