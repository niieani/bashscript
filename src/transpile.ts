import * as ts from 'typescript'
import {translateTsAstToBashAst} from './visitors'
import {write} from './ast/bash-ast'
import {Scope} from './scope'

export function transpile(sourceFile: ts.SourceFile) {
  const {nodes: [bashAst] = [], errors, warnings} = translateTsAstToBashAst({
    tsNodes: [sourceFile],
    scope: new Scope({type: 'FILE', name: sourceFile.fileName}),
  })
  const output = write(bashAst)
  console.log(errors)
  console.log(warnings)
  return output
}

export function transpileCode(code: string, fileName = 'default.ts') {
  const sourceFile: ts.SourceFile = ts.createSourceFile(
    fileName,
    code,
    ts.ScriptTarget.ESNext,
    true,
    ts.ScriptKind.TS,
  )
  return transpile(sourceFile)
}
