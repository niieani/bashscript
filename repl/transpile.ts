import './index.css'
import {makeTestFile} from '../src/util/test-util'
import {transpile} from '../src/transpile'

export const transpileText = (text: string) => {
  const file = makeTestFile(text)
  try {
    return transpile(file)
  } catch {
    return undefined
  }
}
