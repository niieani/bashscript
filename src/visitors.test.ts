import {transpile} from './transpile'
import {makeTestFile} from './util/test-util'

describe('visitors', () => {
  test('variableStatementVisitor', () => {
    const file = makeTestFile(`
      const a = 123
    `)
    expect(transpile(file)).toMatchSnapshot()
  })

  test('importDeclarationVisitor', () => {
    const file = makeTestFile(`
      const a = 123
    `)
    console.log(transpile(file))
  })
})
