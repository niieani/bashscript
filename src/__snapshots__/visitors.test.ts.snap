// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitors variableStatementVisitor 1`] = `
"declare a=# Unsupported initializer type: NumericLiteral

"
`;
